src/rgw/rgw_sal_motr.cc

```diff
@@ -57,6 +61,165 @@ static std::string motr_global_indices[] = {
+	   (left.f_container == right.f_container && left.f_key < right.f_key);
+  }
+};
+static map<struct m0_fid, MotrWatcher*, MotrFidComparator> motr_watchers;
+
+// process_fdmi_record() can't be implemented as MotrWatcher's member function
+// as the function pointer to a member funciton is a different type to fdmi's
+// callback function.
+static int process_fdmi_record(struct m0_uint128 *rec_id,
+                               struct m0_buf fdmi_rec,
+                               struct m0_fid filter_id)
+{
+  int rc;
+  int len;
+  struct m0_fol_rec fol_rec;
+  struct m0_fol_frag *frag;
+  struct m0_fop_fol_frag *fp_frag;
+  struct m0_cas_op *cas_op;
+  struct m0_cas_recv cg_rec;
+  struct m0_cas_rec *cr_rec;
+  char *addr;
+  void *next;
+  const struct m0_tl *head = &fol_rec.fr_frags;
+
+  std::cout << "fdmi cb: process_fdmi_record() enter. " << std::endl;
+
+  auto iter = motr_watchers.find(filter_id);
+  if (iter == motr_watchers.end())
+    return -EINVAL;
+  MotrWatcher* watcher = iter->second;
+
+  m0_fol_rec_init(&fol_rec, NULL);
+  rc = m0_fol_rec_decode(&fol_rec, &fdmi_rec);
+  if (rc != 0)
+    goto out;
+
+  // C++ doesn't allow assigning void* to pointer of other type, so
+  // m0_tl_for can't be used here.
+  // error: invalid conversion from ‘void*’ to ‘m0_fol_frag*’ ...
+  // m0_tl_for(m0_rec_frag, &fol_rec.fr_frags, frag) {
+
+  for (frag = (struct m0_fol_frag *)m0_tlist_head(&m0_rec_frag_tl, head);
+       frag != NULL &&	((void)(next = m0_tlist_next(&m0_rec_frag_tl, head, frag)), true);
+       frag = (struct m0_fol_frag *) next)
+  {
+    fp_frag = (struct m0_fop_fol_frag *) frag->rp_data;
+    cas_op = (struct m0_cas_op *) fp_frag->ffrp_fop;
+    cg_rec = cas_op->cg_rec;
+    cr_rec = cg_rec.cr_rec;
+
+    for (uint64_t i = 0; i < cg_rec.cr_nr; i++) {
+      len  = cr_rec[i].cr_val.u.ab_buf.b_nob;
+      addr = (char *) cr_rec[i].cr_val.u.ab_buf.b_addr;
+      bufferlist bl;
+      bl.append(addr, len);
+      rc = watcher->watch_cb(bl);
+      if (rc < 0)
+        goto out;
+    }
+  } //m0_tl_endfor;
+
+out:
+  m0_fol_rec_fini(&fol_rec);
+  std::cout << "fdmi cb: process_fdmi_record() leave rc =  " << rc << std::endl;
+  return rc;
+}
+
+int MotrWatcher::init_fdmi_plugin(const DoutPrefixProvider *dpp)
+{
+  int rc;
+  this->fdmi_dock_ops = m0_fdmi_plugin_dock_api_get();
+  std::cout << "init_fdmi_plugin " << std::endl;
+
+  const auto& filter_id = g_conf().get_val<std::string>("motr_cache_fdmi_filter_id");
+  rc = m0_fid_sscanf(filter_id.c_str(), &this->fdmi_plugin_fid);
+  if (rc < 0)
+    return rc;
+  this->fdmi_plugin_cb.po_fdmi_rec = process_fdmi_record;
+  const struct m0_fdmi_filter_desc fd;
+  ldpp_dout(dpp, 0) << "register filter " << dendl;
+  rc = this->fdmi_dock_ops->fpo_register_filter(&this->fdmi_plugin_fid, &fd,
+                                                &this->fdmi_plugin_cb);
+  if (rc != 0)
+    return rc;
+
+  ldpp_dout(dpp, 0) << "enable filter " << dendl;
```

Comment:
The logging level should be `20` here, perhaps? And in other similar places where it's not an error case.